### PR TITLE
Enabled CLI to read predefined and customized CSV files

### PR DESCRIPTION
**Pull request by lziq**
*Issue #366*

*Description of changes:*
1. Enabled CLI to read predefined and customized CSV files. 
2. Wrote a few test cases. 
3. Added documentation.

### CODE REVIEW

File: cli/test/org/partiql/cli/functions/ReadFileTest.kt [hunk 1]
```diff
@@ -151,4 +151,64 @@ class ReadFileTest {
...
+
+        val actual = function.call(env, args).ionValue
+        val expected = "[{title:\"harry potter\",category:\"book\",price:\"7.99\"}]"
+
+        assertEquals(ion.singleValue(expected), actual)
+    }
+
+    @Test
+    fun readPostgreCsvFile() {
+        writeFile("simple_postgre.csv", "id,name,balance\n1,Bob,10000.00")
```
Comment (alancai98) [inline cli/test/org/partiql/cli/functions/ReadFileTest.kt:169]: typo: 'simple_postgre.csv' -> 'simple_postgresql.csv'

```suggestion
    fun readPostgreSQLCsvFile() {
        writeFile("simple_postgresql.csv", "id,name,balance\n1,Bob,10000.00")
```

File: cli/test/org/partiql/cli/functions/ReadFileTest.kt [hunk 2]
```diff
@@ -151,4 +151,64 @@ class ReadFileTest {
...
+    fun readCustomizedCsvFile3() { // trim and ignore_surrounding_space
+        writeFile("customized.csv", "id,name,balance\n 1 , Bob , 10000.00 ")
+
+        val args = listOf("\"${dirPath("customized.csv")}\"", "{type:\"customized\", header:true, ignore_surrounding_space:false, trim:false}").map { it.exprValue() }
+
+        val actual = function.call(env, args).ionValue
+        val expected = "[{id:\" 1 \",name:\" Bob \",balance:\" 10000.00 \"}]"
+
+        assertEquals(ion.singleValue(expected), actual)
+    }
```
Comment (alancai98) [inline cli/test/org/partiql/cli/functions/ReadFileTest.kt:213]: Not all the added custom configuration options have been tested here. Would be good to include them or create a separate issue to add them.

Similarly for the other csv `type`s like `mysql_csv`, `mongodb_csv`, etc.

**Review by alancai98 — CHANGES_REQUESTED** [review]

File: examples/build.gradle [hunk 3]
```diff
@@ -29,6 +29,8 @@ dependencies {
     implementation 'com.amazonaws:aws-java-sdk-s3:1.11.554'
     implementation 'com.amazonaws:aws-java-sdk-s3control:1.11.554'
 
+    implementation 'org.apache.commons:commons-csv:1.8'
```
[the remark this answers is not in the document]
Comment (lziq) [reply examples/build.gradle:32]: It's not needed.

File: lang/src/org/partiql/lang/eval/io/DelimitedValues.kt [hunk 4]
```diff
@@ -68,16 +68,11 @@ object DelimitedValues {
...
-                  hasHeader: Boolean,
+                  csvFormat: CSVFormat,
                   conversionMode: ConversionMode): ExprValue {
         val reader = BufferedReader(input)
-        val csvFormat = when (hasHeader){
-            true -> CSVFormat.DEFAULT.withDelimiter(delimiter).withFirstRecordAsHeader()
-            false -> CSVFormat.DEFAULT.withDelimiter(delimiter)
-        }
         val csvParser = CSVParser(reader, csvFormat)
-        val columns: List<String> = csvParser.headerNames // `columns` is an empty list when `hasHeader` is false
```
[the remark this answers is not in the document]
Comment (lziq) [reply lang/src/org/partiql/lang/eval/io/DelimitedValues.kt:80]: We don't have `hasHeader` variable in the function `exprValue` any more.

File: docs/user/CLI.md [hunk 5]
```diff
@@ -641,3 +641,33 @@ Kumo	dog
...
+title category price
+harry_potter book 7.99
+dot electronics 49.99
+echo electronics 99.99
+```
+We can use the following command to read the file:
+```
+read_file('customized.csv', {'type':'customized', 'delimiter':' ', 'header':true})
+```
+The following command explicitly shows all the available options for a standard CSV file:
```
[the remark this answers is not in the document]
Comment (lziq) [reply docs/user/CLI.md:668]: Just means the default CSV format in CSVParser library.

File: cli/src/org/partiql/cli/functions/ReadFile.kt [hunk 6]
```diff
@@ -28,14 +29,30 @@ internal class ReadFile(valueFactory: ExprValueFactory) : BaseFunction(valueFact
...
+        val ignoreEmptyLine = options["ignore_empty_line"]?.booleanValue() ?: true
+        val ignoreSurroundingSpace = options["ignore_surrounding_space"]?.booleanValue() ?: true
+        val trim = options["trim"]?.booleanValue() ?: true
+        val delimiter = options["delimiter"]?.stringValue()?.first() // CSVParser library only accepts a single character as delimiter
+        val record = options["line_breaker"]?.stringValue()
+        val escape = options["escape"]?.stringValue()?.first() // CSVParser library only accepts a single character as escape
+        val quote = options["quote"]?.stringValue()?.first() // CSVParser library only accepts a single character as quote
 
-        val reader = InputStreamReader(input, encoding)
+        val csvFormat = csvFormat.let{ it.withIgnoreEmptyLines(ignoreEmptyLine) }
```
[the remark this answers is not in the document]
Comment (lziq) [reply cli/src/org/partiql/cli/functions/ReadFile.kt:46]: I think there is no better way to get `csvFormt`

[the remark this answers is not in the document]
Comment (lziq) [reply cli/src/org/partiql/cli/functions/ReadFile.kt:46]: I think it is okay.

File: cli/test/org/partiql/cli/functions/ReadFileTest.kt (hunk 2 above)
Comment (lziq) [reply cli/test/org/partiql/cli/functions/ReadFileTest.kt:213]: Will add tests for other custom configuration options in the next commit. 
However, the source of other csv types are a bit hard to get. For example, if we have to have `mysql_csv.txt` file, I think we need to actually use MySQL to generate it.

File: docs/user/CLI.md [hunk 7]
```diff
@@ -641,3 +641,33 @@ Kumo	dog
...
+We can use the following command to read the file:
+```
+read_file('customized.csv', {'type':'customized', 'delimiter':' ', 'header':true})
+```
+The following command explicitly shows all the available options for a standard CSV file: 
+```
+read_file('customized.csv', {'type':'customized', 'delimiter':',', 'header':true, \
+'ignore_empty_line':true, 'ignore_surrounding_space':true, 'trim':true, \ 
+'line_breaker: \n', 'escape':'\', 'quote':'"'})
+```
```
[the remark this answers is not in the document]
Comment (lziq) [reply docs/user/CLI.md]: Applied in the next commit.

**Review by alancai98 — COMMENTED** [review]
Nice work! Just have some minor comments and questions. Otherwise, looks good to me.

File: docs/user/CLI.md [hunk 8]
```diff
@@ -641,3 +641,35 @@ Kumo	dog
...
+read_file('customized.csv', {'type':'customized', 'delimiter':' ', 'header':true})
+```
+All the available options for customized CSV files are shown as following: 
+1. Ignore empty lines: `'ignore_empty_line':true`
+2. Ignore spaces surrounding comma:  `'ignore_surrounding_space':true` 
+3. Trim leading and trailing blanks: `'trim':true` 
+4. Set line breaker (only working with '\r', '\n' and '\r\n'): `'line_breaker: \n'`
+5. Set escape sign (single character only): `'escape':'\'`
+6. Set quote sign (single character only): `'quote':'"'`
+7. Set delimiter sign (single character only): `'delimiter':','`
```
[the remark this answers is not in the document]
Comment (lziq) [reply docs/user/CLI.md:675]: Which other options? I think these are all the options to customize the format of a CSV file.

**Comment by lziq** [comment]
Will create a new issue to add tests to read CSV files in other formats. Other comments were already applied.

File: docs/user/CLI.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (alancai98) [reply docs/user/CLI.md]: From the `CSVFormat` constructor and `CSVFormat.Builder` ([documentation here](https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.Builder.html)), there are about 25 total configuration options like `nullString`, `recordSeparator`, `quoteMode`, `skipHeaderRecord`, etc.

[the remark this answers is not in the document]
Comment (lziq) [reply docs/user/CLI.md]: I think other options are not important to configure CSV format. If we find we need any one of them in the future, we can create a ticket for it and make corresponding enhancement.

**Review by alancai98 — APPROVED** [review]
:shipit:

**Pull request merged**